server/src/main/scala/org/http4s/server/ServerBuildable.scala

```diff
@@ -0,0 +1,141 @@
+ * limitations under the License.
+ */
+
+package org.http4s
+package server
+
+import cats.effect.concurrent._
+import cats.effect._
+import cats.syntax.all._
+import fs2._
+import fs2.concurrent._
+
+/** A pseudo-typeclass for something which can build a server.
+  *
+  * There are primary three motivations for this class.
+  *
+  * - Provide an ad-hoc polymorphic way approach to configure a server, as
+  *   opposed to the [[ServerBuilder]] subtyping approach. This is generally
+  *   more consistent with the current design idioms in use in FP Scala code
+  *   and (subjectively) is cleaner, e.g. no dealing with `Self` typing, etc.
+  * - Provide a minimal definition for what you need to provide if you are
+  *   implementing a server backend. [[ServerBuilder]] already does this to a
+  *   degree, but due to the current design of that class, not all of our
+  *   servers implement that, e.g. Ember.
+  *
+  * This class does not and ''can not'' replace backend specific
+  * builders. Those must always exist as the precise feature set of a given
+  * backend will likely never be totally shared with any other backend.
+  */
+trait ServerBuildable[F[_], A] {
+
+  /** Returns the backend as a resource.  Resource acquire waits
+    * until the backend is ready to process requests.
+    */
+  def resource(a: A): Resource[F, Server]
+
+  // Abstract Settings
+
+  /** Set the http host for the server. */
+  def withHttpHost(a: A)(host: String): A
+
+  /** Set the http port for the server. */
+  def withHttpPort(a: A)(port: Int): A
+
+  /** Set the [[HttpApp]] to serve on this server. */
+  def withHttpApp(a: A)(app: HttpApp[F]): A
+
+  // final
+
+  /** Returns the backend as a single-element stream.  The stream
+    * does not emit until the backend is ready to process requests.
+    * The backend is shut down when the stream is finalized.
+    */
+  final def stream(a: A): Stream[F, Server] = Stream.resource(resource(a))
+
+  /** Returns an effect that allocates a backend and an `F[Unit]` to
+    * release it.  The returned `F` waits until the backend is ready
+    * to process requests.  The second element of the tuple shuts
+    * down the backend when run.
+    *
+    * Unlike [[resource]] and [[stream]], there is no automatic
+    * release of the backend.  This function is intended for REPL
+    * sessions, tests, and other situations where composing a
+    * [[cats.effect.Resource]] or [[fs2.Stream]] is not tenable.
+    * [[resource]] or [[stream]] is recommended wherever possible.
+    */
+  final def allocated(a: A)(implicit F: Bracket[F, Throwable]): F[(Server, F[Unit])] = resource(
+    a).allocated
+
+  /** Bind the server to a given port and host. */
+  final def bindHttp(a: A)(port: Int = defaults.HttpPort, host: String = defaults.Host): A =
+    withHttpPort(withHttpHost(a)(host))(port)
+
+  /** Bind the server to localhost */
+  final def bindLocal(a: A)(port: Int): A = bindHttp(a)(port, defaults.Host)
```

Comment:
If local is the default, I don't think we need this.

Reply:
I'll remove